daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala

```diff
@@ -214,82 +214,44 @@ private[lf] class PackageInterface(signatures: PartialFunction[PackageId, Packag
   ): Either[LookupError, TemplateChoiceSignature] =
     lookupTemplateChoice(tmpName, chName, Reference.TemplateChoice(tmpName, chName))
 
-  private[this] def lookupTemplateImplements(
-      tmpName: TypeConName,
-      ifaceName: TypeConName,
-      context: => Reference,
-  ): Either[LookupError, TemplateImplementsSignature] =
-    lookupTemplate(tmpName, context).flatMap(
-      _.implements
-        .get(ifaceName)
-        .toRight(LookupError(Reference.TemplateImplements(tmpName, ifaceName), context))
-    )
-
-  def lookupTemplateImplements(
-      tmpName: TypeConName,
-      ifaceName: TypeConName,
-  ): Either[LookupError, TemplateImplementsSignature] =
-    lookupTemplateImplements(tmpName, ifaceName, Reference.TemplateImplements(tmpName, ifaceName))
-
-  private[this] def lookupInterfaceCoImplements(
-      tmpName: TypeConName,
-      ifaceName: TypeConName,
-      context: => Reference,
-  ): Either[LookupError, InterfaceCoImplementsSignature] =
-    lookupInterface(ifaceName, context).flatMap(
-      _.coImplements
-        .get(tmpName)
-        .toRight(LookupError(Reference.InterfaceCoImplements(tmpName, ifaceName), context))
-    )
-
-  def lookupInterfaceCoImplements(
-      tmpName: TypeConName,
-      ifaceName: TypeConName,
-  ): Either[LookupError, InterfaceCoImplementsSignature] =
-    lookupInterfaceCoImplements(
-      tmpName,
-      ifaceName,
-      Reference.InterfaceCoImplements(tmpName, ifaceName),
-    )
-
-  private[this] def lookupTemplateImplementsOrInterfaceCoImplements(
-      tmpName: TypeConName,
-      ifaceName: TypeConName,
+  private[this] def lookupInterfaceInstance(
+      interfaceName: TypeConName,
+      templateName: TypeConName,
       context: => Reference,
   ): Either[
-    LookupError,
-    TemplateOrInterface[
-      TemplateImplementsSignature,
-      InterfaceCoImplementsSignature,
-    ],
+    Either[LookupError, AmbiguousInterfaceInstanceError],
```

Comment:
the nested `Either` is because we have two different error conditions: no `interface instance` exists at all, or both possible ones do. We could have folded `AmbiguousInterfaceInstanceError` into `LookupError`, but quite a few places match on `LookupError` so that would affect too much unrelated code.